File: public/code/chartpatterns__inverse_head_shoulders.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


def pivots(c, lb=5):
    sh = swingHighs(c, lb)
    sl = swingLows(c, lb)
    highs = []
    lows = []
    for i in range(0, len(c)):
        if sh[i] != None:
            highs.append({'idx': i, 'price': sh[i]})
        if sl[i] != None:
            lows.append({'idx': i, 'price': sl[i]})
    return {'highs': highs, 'lows': lows}
def inverseHeadShoulders(c):
    if len(c) < 60:
        return make_signal(reason="Insufficient data")
    _d = pivots(c, 4)
    highs = _d['highs']
    lows = _d['lows']
    if len(lows) < 3  or  len(highs) < 2:
        return make_signal(reason="Not enough pivots")
    [ls, head, rs] = lows[-3:]
    a = atr(c, 14)
    i = len(c) - 1
    cur = c[i]
    if not (head.price < ls.price  and  head.price < rs.price):
        return make_signal(reason="No iH&S structure")
    if abs(ls.price - rs.price) / ls.price > 0.03:
        return make_signal(reason="Shoulders uneven")
    neckHighs = [h for h in highs if h.idx > ls.idx  and  h.idx < rs.idx]
    if len(neckHighs) < 1:
        return make_signal(reason="No neckline")
    neckline = sum(h.price for h in neckHighs) / len(neckHighs)
    if cur['close'] > neckline  and  c[i - 1]['close'] <= neckline:
        sl = rs.price - 0.5 * a[i]
        height = neckline - head.price
        return make_signal(signal="long", entry=cur['close'], stop_loss=sl, take_profit=[cur['close'] + height * 0.5, cur['close'] + height, cur['close'] + height * 1.5], confidence=0.76, reason="Inverse H&S neckline break")
    return make_signal(reason="iH&S forming")

File: public/code/chartpatterns__inverse_head_shoulders.ts
import { Candle, Signal, makeSignal, atr, sma, swingHighs, swingLows } from "../indicators";

function pivots(c: Candle[], lb = 5) {
  const sh = swingHighs(c, lb), sl = swingLows(c, lb);
  const highs: { idx: number; price: number }[] = [], lows: { idx: number; price: number }[] = [];
  for (let i = 0; i < c.length; i++) {
    if (sh[i] !== null) highs.push({ idx: i, price: sh[i] as number });
    if (sl[i] !== null) lows.push({ idx: i, price: sl[i] as number });
  }
  return { highs, lows };
}

export function inverseHeadShoulders(c: Candle[]): Signal {
  if (c.length < 60) return makeSignal({ reason: "Insufficient data" });
  const { highs, lows } = pivots(c, 4);
  if (lows.length < 3 || highs.length < 2) return makeSignal({ reason: "Not enough pivots" });
  const [ls, head, rs] = lows.slice(-3);
  const a = atr(c, 14), i = c.length - 1, cur = c[i];
  if (!(head.price < ls.price && head.price < rs.price)) return makeSignal({ reason: "No iH&S structure" });
  if (Math.abs(ls.price - rs.price) / ls.price > 0.03) return makeSignal({ reason: "Shoulders uneven" });
  const neckHighs = highs.filter((h) => h.idx > ls.idx && h.idx < rs.idx);
  if (neckHighs.length < 1) return makeSignal({ reason: "No neckline" });
  const neckline = neckHighs.reduce((s, h) => s + h.price, 0) / neckHighs.length;
  if (cur.close > neckline && c[i - 1].close <= neckline) {
    const sl = rs.price - 0.5 * a[i], height = neckline - head.price;
    return makeSignal({ signal: "long", entry: cur.close, stop_loss: sl, take_profit: [cur.close + height * 0.5, cur.close + height, cur.close + height * 1.5], confidence: 0.76, reason: "Inverse H&S neckline break" });
  }
  return makeSignal({ reason: "iH&S forming" });
}
